docs/environment_convergence.md

```diff
@@ -0,0 +1,64 @@
+# Environment Convergence
+
+The term "environment" as used in this document refers to a directory on the server, such as `/etc/puppetlabs/code/environments/<name>`, containing puppet manifests, hiera data, custom facts, etc.
+
+At the beginning of an agent run, the agent and server negotiate which environment to use. It is important for the agent and server to use the same environment during the run, because the manifest may reference facts that must be downloaded and evaluated on the agent. If they are mismatched, then compilation can fail.
+
+For security reasons, puppet defaults to **server-specified environments**. This means the server always decides which environment to assign to each agent. This is important because we don't want an agent to request a catalog from an arbitrary environment, as the server might include a class containing sensitive data.
+
+However, there are cases during [iterative development](https://puppet.com/docs/pe/latest/environment_based_testing.html) where it is necessary for a trusted user to be able to run the agent against an environment contained in a feature branch. For example, `puppet agent -t --environment <feature>`. If the code works as intended, then it provides confidence that the feature branch code can be merged. We refer to this workflow as an **agent-specified environment**.
+
+## Agent and Server Negotiation
+
+The server decides which environment an agent should use by asking its node terminus to classify the node.Terminus just means there's an implementation of an interface that knows how to perform CRUD operations on `Puppet::Node` objects. The terminus is responsible for returning a node object, including the environment that the node is assigned to.
+
+PE ships with a `classifier` node terminus that communicates with the PE classifier via REST. So by default, PE uses server-specified environments.
+
+However, open source defaults to a `plain` node terminus, which means it falls back to agent-specified environments.
+
+When the agent negotiates with the server, the server may respond in one of the following ways:
+
+  1. Tell the agent to switch to the server-specified environment.
+  1. Allow the agent to continue using its current agent-specified environment.
+  1. Fail the run due to a classification conflict. Conflicts can arise if the agent is assigned to multiple environments at the same time, as can happen if the agent's facts match multiple rules in the classifier.
+  1. Fail the request because the agent's requested environment doesn't exist on the server.
```

Comment:
I'm not sure if this is a mark down issue or not, but this numbering looks odd.

Reply:
I thought this was a standard markdown thing, but maybe it's a github flavored markdown thing so it autoincrements. In any case, I changed it to 1., 2., etc